File: 백준/Gold/11559. Puyo Puyo/Puyo Puyo.java
import java.io.*;
import java.util.*;
public class Main{
    static char[][] field = new char[12][6];
    static int[] dx = {0,0,-1,1}, dy = {1,-1,0,0};
    static boolean[][] visit;
    static boolean check = false;

    static void bfs(int x,int y, char color){
        Queue<int[]> q1 = new LinkedList<>();
        Queue<int[]> q2 = new LinkedList<>();

        q1.add(new int[]{x,y});
        q2.add(new int[]{x,y});
        int res = 1;
        while(!q1.isEmpty()){
            int[] cur = q1.poll();
            for (int d = 0; d < 4; d++) {
                int nx = cur[0] + dx[d], ny = cur[1] + dy[d];
                if (nx<0 || ny<0 || nx>=12 || ny>=6 || visit[nx][ny] || field[nx][ny]!=color) continue;
                visit[nx][ny] = true;
                res++;
                q1.offer(new int[]{nx,ny});
                q2.offer(new int[]{nx,ny});
            }
        }

        if (res<4) return;
        check = true;
        while(!q2.isEmpty()){
            int[] cur = q2.poll();
            field[cur[0]][cur[1]] = '.';
        }
    }

    static void down(){
        for (int i = 0; i < 6; i++) {
            boolean c = false;
            int idx = 0;
            for (int j = 11; j >= 0; j--) {
                if (!c){
                    if (field[j][i]=='.') {
                        c = true;
                        idx = j;
                    }
                }
                else {
                    if (field[j][i]!='.') {
                        field[idx][i] = field[j][i];
                        field[j][i] = '.';
                        idx--;
                    }
                }
            }
        }
    }


    public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//        StringBuilder ans = new StringBuilder();
        int ans =0;
        for (int i = 0; i < 12; i++) field[i] = br.readLine().toCharArray();
        while (true) {
            visit = new boolean[12][6];
            for (int i = 0; i < 12; i++) {
                for (int j = 0; j < 6; j++) {
                    if (visit[i][j] || field[i][j]=='.') continue;
                    visit[i][j] = true;
                    bfs(i,j,field[i][j]);
                }
            }
            if (!check) break;
            ans++;
            check = false;
            down();
        }
        System.out.println(ans);
    }
}

File: 백준/Gold/11559. Puyo Puyo/Puyo Puyo.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

//var in = bufio.NewScanner(os.Stdin)
//
//func nextInt() int {
//	in.Scan()
//	r := 0
//	for _, c := range in.Bytes() {
//		r *= 10
//		r += int(c - '0')
//	}
//	return r
//}

const (
	r = 12
	c = 6
)

var (
	field [r][c]rune
	visit [r][c]bool
	dx    = [4]int{0, 0, -1, 1}
	dy    = [4]int{-1, 1, 0, 0}
)

func bfs(x, y int, color rune) bool {
	q := [][2]int{{x, y}}
	toClear := [][2]int{{x, y}}
	cnt := 1
	visit[x][y] = true

	for len(q) > 0 {
		cur := q[0]
		q = q[1:]
		for d := 0; d < 4; d++ {
			nx, ny := cur[0]+dx[d], cur[1]+dy[d]
			if nx < 0 || ny < 0 || nx >= r || ny >= c || visit[nx][ny] || field[nx][ny] != color {
				continue
			}
			visit[nx][ny] = true
			q = append(q, [2]int{nx, ny})
			toClear = append(toClear, [2]int{nx, ny})
			cnt++
		}
	}
	if cnt >= 4 {
		for _, cur := range toClear {
			field[cur[0]][cur[1]] = '.'
		}
		return true
	}
	return false
}

func down() {
	for j := 0; j < c; j++ {
		c := false
		idx := 0
		for i := r - 1; i >= 0; i-- {
			if !c {
				if field[i][j] == '.' {
					c = true
					idx = i
				}
			} else {
				if field[i][j] != '.' {
					field[idx][j] = field[i][j]
					field[i][j] = '.'
					idx--
				}
			}
		}
	}
}

func main() {
	br := bufio.NewReader(os.Stdin)
	ans := 0

	for i := range field {
		var line string
		fmt.Fscan(br, &line)
		for j, ch := range line {
			field[i][j] = ch
		}
	}

	for {
		visit = [r][c]bool{}
		hasCleared := false
		for i := 0; i < r; i++ {
			for j := 0; j < c; j++ {
				if visit[i][j] || field[i][j] == '.' {
					continue
				}
				if bfs(i, j, field[i][j]) {
					hasCleared = true
				}

			}
		}
		if !hasCleared {
			break
		}
		ans++
		down()
	}

	fmt.Println(ans)
}
